src/box/lua/config/applier/credentials.lua

```diff
@@ -277,85 +282,291 @@ local function privileges_add_defaults(name, role_or_user, intermediate)
--- 3. Revoke all privileges that are not present in B,
---    but present in A (`revoke(A - B)).
---
--- Default privileges are not granted on step 1, so they stay revoked if
--- revoked manually (e.g. with `box.schema.{user,role}.revoke()`).
--- However, defaults should never be revoked, so target state B is enriched
--- with them before step 3.
-local function sync_privileges(name, config_privileges, role_or_user)
-    assert(role_or_user == 'user' or role_or_user == 'role')
-    log.verbose('syncing privileges for %s %q', role_or_user, name)
+-- Get the latest credentials configuration from the config and add empty
+-- default users and roles configuration, if they are missing.
+local function get_credentials(config)
+    local configdata = config._configdata
+    local credentials = configdata:get('credentials')
+
+    -- If credentials section in config is empty, skip applier.
+    if credentials == nil then
+        return {}
+    end
+
+    -- Tarantool has the following roles and users present by default on every
+    -- instance:
+    --
+    -- Default roles:
+    -- * super
+    -- * public
+    -- * replication
+    --
+    -- Default users:
+    -- * guest
+    -- * admin
+    --
+    -- These roles and users have according privileges pre-granted by design.
+    -- Credentials applier adds such privileges with `priviliges_add_default()`
+    -- when syncing. So, for the excessive (non-default) privs to be removed,
+    -- these roles and users must be present inside configuration at least in a
+    -- form of an empty table. Otherwise, the privileges will be left unchanged,
+    -- similar to all used-defined roles and users.
+
+    credentials.roles = credentials.roles or {}
+    credentials.roles['super'] = credentials.roles['super'] or {}
+    credentials.roles['public'] = credentials.roles['public'] or {}
+    credentials.roles['replication'] = credentials.roles['replication'] or {}
+
+    credentials.users = credentials.users or {}
+    credentials.users['guest'] = credentials.users['guest'] or {}
+    credentials.users['admin'] = credentials.users['admin'] or {}
+
+    return credentials
+end
+
+-- }}} Sync helpers
+
+-- {{{ Triggers for grants after obj creation
+
+-- This is fiber channel that contains all scheduled tasks
+-- in form of a table with the following format:
+-- {type = 'BLOCKING_FULL_SYNC'/'BACKGROUND_FULL_SYNC'/obj_type,
+--  name = obj_name}
+local sync_tasks
 
-    local grant_f = function(name, privs, obj_type, obj_name)
-        privs = table.concat(privs, ',')
-        log.debug('credentials.apply: ' .. role_or_user ..
-                  '.grant(%q, %q, %q, %q)', name, privs, obj_type, obj_name)
-        box.schema[role_or_user].grant(name, privs, obj_type, obj_name)
+-- The table contains flags for objects that should be synced in the trigger:
+-- [obj_type][obj_name] = true/nil
+local on_creation_flags = {
+    ['space'] = {},
+    ['function'] = {},
+    ['sequence'] = {},
+}
```

Comment:
I think that it would be helpful to highlight that it is a mapping from a current config. It worth to reflect it in the name and it would be good to mention that it is always in sync with a latest config in a comment.

Reply:
Updated:
```lua
  -- This is a map of all objects that are mentioned in the config.
  -- It is always in sync with the latest reload and has the following format:
  -- [obj_type][obj_name] = true/nil                                                                     
  local target_object_map = {                                                                         
      ['space'] = {},                                                                                    
      ['function'] = {},                                                                                 
      ['sequence'] = {},                                                                                 
  }                                                                                                      
```